quiche/src/lib.rs

```diff
@@ -8388,13 +8554,144 @@ impl TransportParams {
                 ),
                 initial_max_streams_bidi: Some(self.initial_max_streams_bidi),
                 initial_max_streams_uni: Some(self.initial_max_streams_uni),
-
-                preferred_address: None,
+                preferred_address,
             },
         )
     }
 }
 
+/// Preferred Address Transport Parameters
+#[derive(Clone, Debug, PartialEq)]
+pub struct PreferredAddressParams {
+    /// The server's preferred V4 address
+    pub addr_v4: Option<SocketAddrV4>,
+    /// The server's preferred V6 address
+    pub addr_v6: Option<SocketAddrV6>,
+    /// The preferred address connection ID
+    pub connection_id: ConnectionId<'static>,
+    /// The preferred address reset token
+    pub stateless_reset_token: u128,
+}
+
+impl PreferredAddressParams {
+    fn new(
+        addr_v4: Option<SocketAddrV4>, addr_v6: Option<SocketAddrV6>,
+        connection_id: ConnectionId<'static>, stateless_reset_token: u128,
+    ) -> Self {
+        PreferredAddressParams {
+            addr_v4,
+            addr_v6,
+            connection_id,
+            stateless_reset_token,
+        }
+    }
+
+    fn decode(b: &mut octets::Octets) -> Result<Self> {
+        let addr_v4 = PreferredAddressParams::decode_v4(b)?;
+        let addr_v6 = PreferredAddressParams::decode_v6(b)?;
+
+        if addr_v4.is_none() && addr_v6.is_none() {
+            return Err(Error::InvalidTransportParam);
+        }
+
+        let cid = b.get_bytes_with_varint_length()?.to_vec().into();
+        let stateless_reset_token = u128::from_be_bytes(
+            b.get_bytes(16)?.to_vec().try_into().unwrap(), /* Since a 128 bit
+                                                            * vector has just
+                                                            * successfully
+                                                            * been created
+                                                            * above, this
+                                                            * should never
+                                                            * panic. */
+        );
+
+        Ok(PreferredAddressParams::new(
+            addr_v4,
+            addr_v6,
+            cid,
+            stateless_reset_token,
+        ))
+    }
+
+    fn decode_v4(b: &mut octets::Octets) -> Result<Option<SocketAddrV4>> {
+        let mut addr_v4 = None;
+        let ip_v4_b = <[u8; 4]>::try_from(b.slice(4)?)
+            .map_err(|_| Error::InvalidTransportParam)?;
+        b.skip(4)?;
+        let ip_v4 = Ipv4Addr::from(ip_v4_b);
+
+        let port_v4 = b.get_u16()?;
+
+        if ip_v4.is_unspecified() ^ (port_v4 == 0) {
+            return Err(Error::InvalidTransportParam);
```

Comment:
I think this is trying to enforce https://datatracker.ietf.org/doc/html/rfc9000#section-18.2-4.32.1 ?

> Servers MAY choose to only send a preferred address of one address family by sending an all-zero address and port (0.0.0.0:0 or [::]:0) for the other family. IP addresses are encoded in network byte order.

My concern is that this is called in the top-level `[decode()](https://github.com/cloudflare/quiche/pull/1634/files#diff-2ac678416f8d7ea462a714681b63acedcbbb7f2c99e537ce61b7667b3bf4ac41R8589)` function and would early abort something that is not technicall invalid

Reply:
If the IPv4 address is unspecified (all 0s) and the associated port is also unspecified (all 0s) then the XOR operation will fail and this `decode_v4` function will return Ok(None)